jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DoSHandler.java

```diff
@@ -0,0 +1,792 @@
+            }
+
+            @Override
+            protected boolean onExpired(Tracker tracker)
+            {
+                return tracker.isIdle(NanoTime.now());
+            }
+        };
+        addBean(_cyclicTimeouts);
+        super.doStart();
+    }
+
+    @Override
+    protected void doStop() throws Exception
+    {
+        super.doStop();
+        removeBean(_cyclicTimeouts);
+        _cyclicTimeouts.destroy();
+        _cyclicTimeouts = null;
+    }
+
+    /**
+     * A RateTracker is associated with an id, and stores request rate data.
+     */
+    public abstract static class Tracker implements CyclicTimeouts.Expirable
+    {
+        private final String _id;
+
+        protected Tracker(String id)
+        {
+            _id = id;
+        }
+
+        public String getId()
+        {
+            return _id;
+        }
+
+        public abstract boolean isIdle(long now);
+
+        /**
+         * Add a request to the tracker and check the rate limit
+         *
+         * @param now The timestamp of the request
+         * @return {@code true} if the request is below the limit
+         */
+        public abstract boolean onRequest(long now);
+
+        public abstract int getRequestsPerSecond(long now);
+
+        public String toString()
+        {
+            return "Tracker(%s)@%x".formatted(_id, hashCode());
+        }
+
+        public abstract static class Factory
+        {
+            private static final long MAX_PERIOD_MS = TimeUnit.HOURS.toMillis(1);
+            private static final long ENCODE_NANOS_FACTOR = MAX_PERIOD_MS * 10;
+
+            protected final int _maxRequestsPerSecond;
+            protected final Duration _samplePeriod;
+            protected final int _requestsPerSample;
+
+            public Factory(
+                @Name("maxRequestsPerSecond") int maxRequestsPerSecond,
+                @Name("samplePeriod") Duration samplePeriod)
+            {
+                _maxRequestsPerSecond = maxRequestsPerSecond;
+                _samplePeriod = samplePeriod == null ? Duration.ofSeconds(1) : samplePeriod;
+                if (_samplePeriod.toMillis() >= MAX_PERIOD_MS)
+                    throw new IllegalArgumentException("Sample Period must be less that 1 hour");
+
+                double samplesPerSecond = Duration.ofSeconds(1).toNanos() * 1.0D / _samplePeriod.toNanos();
+                _requestsPerSample = (int)(1.0D * maxRequestsPerSecond / samplesPerSecond);
+            }
+
+            protected static int encodeNanoTime(long nanoTime)
+            {
+                return (int)(TimeUnit.NANOSECONDS.toMillis(nanoTime) % (ENCODE_NANOS_FACTOR));
+            }
+
+            protected static long decodeNanoTime(int encodedTime, long referenceTime)
+            {
+                return TimeUnit.MILLISECONDS.toNanos((TimeUnit.NANOSECONDS.toMillis(referenceTime) / (ENCODE_NANOS_FACTOR)) * ENCODE_NANOS_FACTOR + encodedTime);
+            }
```

Comment:
Is not this hardcoding an implementation into the generic interface to create Trackers?